### PR TITLE
Added support to create Database for a service

### PR DESCRIPTION
**Pull request by shahpratikr**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -78,3 +80,12 @@ Add service URL to the keystone and also generate password to interact with keys
...
 The service-name is obvious, the ingress-suffix is to allow for flexibility to advertise additional suffix for example if you want to advertise qbert/v2 as the suffix instead of just qbert.
+
+## create-db
+Self explanatory
+
+```
+./go-firkinize create-db --service-name hagrid
+{"level":"info","ts":1612943409.0324764,"caller":"cfg/cfgmgr.go:202","msg":"Created DB successfully"}
+{"level":"info","ts":1612943409.0376244,"caller":"cfg/cfgmgr.go:225","msg":"Granted all privileges to DB"}
+```
```
Comment (roopakparikh) [inline README.md:91]: you should print out the name as the output

File: cmd/createdb.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+package cmd
+
+import (
+    "github.com/spf13/cobra"
+    "go.uber.org/zap"
+)
+
+func init() {
+    createDBCmd.PersistentFlags().StringVar(&serviceName, "service-name", "", "Name of the service (example qbert)")
```
Comment (roopakparikh) [inline cmd/createdb.go:16]: where is the variable serviceName declared ?

File: cmd/createdb.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+var createDBCmd = &cobra.Command{
+    Use:   "create-db",
+    Short: "Create DB for service",
+    Long: `Create DB for service`,
+    RunE: func(cmd *cobra.Command, args []string) error {
+        cfgMgr := GetCfg()
+        zap.L().Debug("Creating DB for service")
+        err := cfgMgr.CreateDB(serviceName)
+        if err != nil {
+            zap.L().Info("Error creating DB")
```
Comment (roopakparikh) [inline cmd/createdb.go]: L().Error and also print the "err"

File: cmd/createdb.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            return err
+        }
+        return nil
```
Comment (roopakparikh) [inline cmd/createdb.go:37]: I would print the database name created as the output

File: cmd/root.go [hunk 5]
```diff
@@ -3,7 +3,7 @@
 package cmd
```
Comment (roopakparikh) [inline cmd/root.go:3]: Please add the copyright

File: pkg/cfg/cfgmgr.go [hunk 6]
```diff
@@ -94,15 +97,14 @@ func (c *CfgMgr) GetKeystonePassword(serviceName string) (string, error) {
...
-
 }
 
 //  AddKeystoneUser
 func (c *CfgMgr) AddKeystoneUser(serviceName string) error {
 	zap.L().Debug("Creating keystone user for serviceName ", zap.String("serviceName", serviceName))
 
 	globalKSUserPrefix := fmt.Sprintf("%s/keystone/users/%s", c.CustomerKeyPrefix, serviceName)
-	regionKSUserPrefix := fmt.Sprintf("%s/services/%s/keystne_user/", c.RegionKeyPrefix, serviceName)
+	regionKSUserPrefix := fmt.Sprintf("%s/services/%s/keystone_user", c.RegionKeyPrefix, serviceName)
```
Comment (roopakparikh) [inline pkg/cfg/cfgmgr.go:107]: thanks for fixing

File: pkg/cfg/cfgmgr.go [hunk 7]
```diff
@@ -147,16 +149,87 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
-	out := make([]byte, max)
-	for i := 0; i < max; i++ {
-		out[i] = allChars[rand.Intn(len(allChars))]
-	}
-	return string(out)
+func (c *CfgMgr) CreateDB(serviceName string) error {
+    zap.L().Debug("Creating DB for serviceName ", zap.String("serviceName", serviceName))
+    dbserver, err := c.getValue(fmt.Sprintf("%s/keystone/dbserver_key", c.CustomerKeyPrefix))
+    if err != nil {
+        return err
```
Comment (roopakparikh) [inline pkg/cfg/cfgmgr.go:171]: please print log before you return error

File: pkg/cfg/cfgmgr.go [hunk 8]
```diff
@@ -147,16 +149,87 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+    }
+    host, err := c.getValue(fmt.Sprintf("%s/host", dbserver))
+    if err != nil {
+        return err
```
Comment (roopakparikh) [inline pkg/cfg/cfgmgr.go:176]: here and other places as well

File: cmd/createdb.go (hunk 2 above)
Comment (roopakparikh) [inline cmd/createdb.go:16]: what about user name, isn't that needed ?

File: pkg/cfg/cfgmgr.go [hunk 9]
```diff
@@ -147,16 +149,87 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+    adminPass, err := c.getValue(fmt.Sprintf("%s/admin_pass", dbserver))
+    if err != nil {
+        return err
+    }
+    dbObject, err := sql.Open("mysql", fmt.Sprintf("%s:%s@tcp(%s:%s)/", adminUser, adminPass, host, port))
+    if err != nil {
+        zap.L().Error("Can't connect to MySQL")
+        return err
+    }
+    _, err = dbObject.Exec(fmt.Sprintf("CREATE DATABASE IF NOT EXISTS %s", serviceName))
```
Comment (roopakparikh) [inline pkg/cfg/cfgmgr.go]: What if the database already existed ? We should deal with that too.

File: pkg/cfg/cfgmgr.go [hunk 10]
```diff
@@ -147,16 +149,87 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+    zap.L().Info("Created DB successfully")
+    rows, err := dbObject.Query("SELECT @@hostname")
+    if err != nil {
+        zap.L().Error("Error while getting hostname")
+        return err
+    }
+    defer rows.Close()
+    var hostname string
+    for rows.Next() {
+        _ = rows.Scan(&hostname)
```
Comment (roopakparikh) [inline pkg/cfg/cfgmgr.go:217]: Why is this needed ? Dont we have the hostname already ?

File: pkg/cfg/cfgmgr.go [hunk 11]
```diff
@@ -147,16 +149,87 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+    }
+    dbPassword := c.getRandomPassword()
+    hosts := []string{"localhost", "%", hostname}
+    for _, host := range hosts {
+        query := fmt.Sprintf("GRANT ALL PRIVILEGES ON %s.* TO '%s'@'%s' IDENTIFIED BY '%s'",
```
Comment (roopakparikh) [inline pkg/cfg/cfgmgr.go:222]: Similar to my comment above, can you check if this statement is idempotent

File: cmd/createdb.go (hunk 2 above)
Comment (shahpratikr) [reply cmd/createdb.go:16]: in addkeystone.go

File: cmd/createdb.go (hunk 4 above)
Comment (shahpratikr) [reply cmd/createdb.go:37]: it's already getting printed in cfgmgr.go, do we need it here as well?

File: cmd/root.go (hunk 5 above)
Comment (shahpratikr) [reply cmd/root.go:3]: it's already present in root.go

File: cmd/createdb.go (hunk 2 above)
Comment (shahpratikr) [reply cmd/createdb.go:16]: why username is needed? can you share purpose for it?

File: pkg/cfg/cfgmgr.go (hunk 10 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:217]: where do we get mysql's hostname?

File: pkg/cfg/cfgmgr.go (hunk 11 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:222]: this is to grant all privileges to created DB. without this, you can't access DB in mysql pod.

File: cmd/createdb.go (hunk 2 above)
Comment (poojaghumre) [reply cmd/createdb.go:16]: I had the same question.. for hagrid db creation, we should take db user as input

File: cmd/root.go [hunk 12]
```diff
@@ -63,7 +63,7 @@ func setupLogs() {
 	if debugLog {
 		config = zap.NewDevelopmentConfig()
 	} else {
-		config = zap.NewProductionConfig()				
+		config = zap.NewProductionConfig()
 	}
```
Comment (poojaghumre) [inline cmd/root.go:67]: typos in init func below (cosul -> consul)

File: pkg/cfg/cfgmgr.go [hunk 13]
```diff
@@ -147,16 +149,87 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+    if err != nil {
+        zap.L().Error("Can't connect to MySQL")
+        return err
+    }
+    _, err = dbObject.Exec(fmt.Sprintf("CREATE DATABASE IF NOT EXISTS %s", serviceName))
+    if err != nil {
+        zap.L().Error("Error while creating database")
+        return err
+    }
+    zap.L().Info("Created DB successfully")
```
Comment (poojaghumre) [inline pkg/cfg/cfgmgr.go]: nit: log serviceName

File: pkg/cfg/cfgmgr.go [hunk 14]
```diff
@@ -147,16 +149,87 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+    rows, err := dbObject.Query("SELECT @@hostname")
```
Comment (poojaghumre) [inline pkg/cfg/cfgmgr.go:209]: why is this needed? isn't this same as value in "dbserver_key/host"?

File: pkg/cfg/cfgmgr.go [hunk 15]
```diff
@@ -147,16 +149,87 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+    defer rows.Close()
+    var hostname string
+    for rows.Next() {
+        _ = rows.Scan(&hostname)
+    }
+    dbPassword := c.getRandomPassword()
+    hosts := []string{"localhost", "%", hostname}
+    for _, host := range hosts {
+        query := fmt.Sprintf("GRANT ALL PRIVILEGES ON %s.* TO '%s'@'%s' IDENTIFIED BY '%s'",
+                             serviceName, serviceName, host, dbPassword)
```
Comment (poojaghumre) [inline pkg/cfg/cfgmgr.go]: take userName as an argument (you can keep it optional, and default to serviceName)

**Review by poojaghumre — CHANGES_REQUESTED** [review]
Thanks Pratik! some minor comments..

File: pkg/cfg/cfgmgr.go (hunk 14 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:209]: no. this is different value. something like "mysql_*". dbserver_key/host returns 127.0.0.1

File: cmd/createdb.go (hunk 4 above)
Comment (roopakparikh) [reply cmd/createdb.go:37]: I don't see it . As a general rule make sure that we use pkg to act like a library and not to interact or print at the cmd level.

File: pkg/cfg/cfgmgr.go (hunk 7 above)
Comment (roopakparikh) [reply pkg/cfg/cfgmgr.go:171]: Indentation of the zap.L() is incorrect.

File: pkg/cfg/cfgmgr.go [hunk 16]
```diff
@@ -147,16 +149,97 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+func (c *CfgMgr) CreateDB(serviceName, userName string) error {
+    zap.L().Debug("Creating DB for serviceName ", zap.String("serviceName", serviceName))
+    dbserver, err := c.getValue(fmt.Sprintf("%s/keystone/dbserver_key", c.CustomerKeyPrefix))
+    if err != nil {
+    	zap.L().Error("Cannot get dbserver_key from consul store", zap.Error(err))
+        return err
+    }
+    host, err := c.getValue(fmt.Sprintf("%s/host", dbserver))
+    if err != nil {
+    	zap.L().Error("Cannot get host key from consul store", zap.Error(err))
```
Comment (roopakparikh) [inline pkg/cfg/cfgmgr.go]: Same issue in other places, please fix it before you merge

File: pkg/cfg/cfgmgr.go (hunk 10 above)
Comment (roopakparikh) [reply pkg/cfg/cfgmgr.go:217]: what is the variable "host" ?

File: pkg/cfg/cfgmgr.go (hunk 11 above)
Comment (roopakparikh) [reply pkg/cfg/cfgmgr.go:222]: The question is: can you run this statement again and again, in other words should we check if the privileges are already granted ? Otherwise if you run the script again the second time it will unnecessary fail

File: pkg/cfg/cfgmgr.go [hunk 17]
```diff
@@ -147,16 +149,97 @@ func (c *CfgMgr) getValue(key string) (string, error) {
...
+    if err == nil {
+    	zap.L().Info(fmt.Sprintf("Database %s already exists", dbName))
+    	return nil
+    }
+    dbObject, err := sql.Open("mysql", fmt.Sprintf("%s:%s@tcp(%s:%s)/", adminUser, adminPass, host, port))
+    if err != nil {
+        zap.L().Error("Can't connect to MySQL")
+        return err
+    }
+    _, err = dbObject.Exec(fmt.Sprintf("CREATE DATABASE %s", serviceName))
```
Comment (roopakparikh) [inline pkg/cfg/cfgmgr.go:203]: Like my other comment, can you make sure if we execute the createdb command again it will not fail. In other words will create database fail if the database already exists, if it does please ignore that specific error or check if the database already exists and not call create

**Review by roopakparikh — COMMENTED** [review]
Pratik, a few minor comments, but the most important one is the 'idempotency' can we call create-db again and again without it failing the second or third time

File: pkg/cfg/cfgmgr.go (hunk 17 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:203]: `mysql> select @@hostname;
+------------------------+
| @@hostname             |
+------------------------+
| mysql-7f4fcc7b87-wqmsl |
+------------------------+`
we capture this value in hostname

Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:203]: we are checking if db is present or not at https://github.com/platform9/go-firkinize/pull/1/files#diff-f029d585806a2c597e25fe0422e2db28b6aed18e54b3c1962c8cef5bfb28ec67R193

Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:203]: `# ./root/go-firkinize create-db --service-name=hagrid --consul-token=88fc02fe-0034-93a1-6a9e-e56eaedf9cb9 --region-id=5aaaff0a-a48e-491d-8b8d-5c886fe6ae73 --customer-id=15ab505e-ba46-4cd5-a149-65afff82085d --consul-host-port="konsul-dev.platform9.horse:443" --consul-scheme="https"
{"level":"info","ts":1613116062.0443702,"caller":"cfg/cfgmgr.go:195","msg":"Database hagrid already exists"}`
if we run again and if DB already exists, then it will give above message and exit with 0 status

File: pkg/cfg/cfgmgr.go (hunk 10 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:217]:
```
mysql> select @@hostname; 
+------------------------+
| @@hostname             |
+------------------------+
| mysql-7f4fcc7b87-wqmsl |
+------------------------+
````
we capture this value in hostname

File: pkg/cfg/cfgmgr.go (hunk 11 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:222]: this statement will get execute only once while creating DB. if DB exists, we are returning from this function on line 193

File: pkg/cfg/cfgmgr.go (hunk 14 above)
Comment (shahpratikr) [reply pkg/cfg/cfgmgr.go:209]: https://github.com/platform9/go-firkinize/pull/1/files?file-filters%5B%5D=.go#r575040392

**Review by roopakparikh — APPROVED** [review]

**Review by poojaghumre — APPROVED** [review]
Thanks for making the changes, Pratik!
